pkg/object/validator/validate.go

```diff
@@ -0,0 +1,186 @@
+)
+
+// Validator contains functionality for validating a set of resources prior
+// to being used by the Apply functionality. This imposes some constraint not
+// always required, such as namespaced resources must have the namespace set.
+type Validator struct {
+	Mapper meta.RESTMapper
+}
+
+// Validate validates the provided resources. A RESTMapper will be used
+// to fetch type information from the live cluster.
+func (v *Validator) Validate(objs []*unstructured.Unstructured) *MultiValidationError {
+	crds := findCRDs(objs)
+	var allErrors []*ValidationError
+
+	ids := UnstructuredSetToObjMetadataSet(objs)
+
+	for i, obj := range objs {
+		id := ids[i]
+		var objErrors []error
+		if err := v.validateName(obj); err != nil {
+			objErrors = append(objErrors, err)
+		}
+		if err := v.validateNamespace(obj, crds); err != nil {
+			objErrors = append(objErrors, err)
+		}
+		if errs := v.validateDependsOnAnnotation(obj, ids); len(errs) > 0 {
+			objErrors = append(objErrors, errs...)
+		}
+		if errs := v.validateMutationAnnotation(obj, ids); len(errs) > 0 {
+			objErrors = append(objErrors, errs...)
+		}
+		if len(objErrors) > 0 {
+			allErrors = append(allErrors, &ValidationError{
+				Identifier: id,
+				Errors:     objErrors,
+			})
+		}
+	}
+	if len(allErrors) > 0 {
+		return &MultiValidationError{
+			Errors: allErrors,
+		}
+	}
+	return nil
+}
+
+// findCRDs looks through the provided resources and returns a slice with
+// the resources that are CRDs.
+func findCRDs(us []*unstructured.Unstructured) []*unstructured.Unstructured {
+	var crds []*unstructured.Unstructured
+	for _, u := range us {
+		if object.IsCRD(u) {
+			crds = append(crds, u)
+		}
+	}
+	return crds
+}
+
+// validateName validates the value of the name field of the resource.
+func (v *Validator) validateName(u *unstructured.Unstructured) error {
+	if u.GetName() == "" {
+		return field.Required(field.NewPath("metadata", "name"), "name is required")
+	}
+	return nil
+}
+
+// validateNamespace validates the value of the namespace field of the resource.
+func (v *Validator) validateNamespace(u *unstructured.Unstructured, crds []*unstructured.Unstructured) error {
+	scope, err := object.LookupResourceScope(u, crds, v.Mapper)
+	if err != nil {
+		return err
+	}
+
+	ns := u.GetNamespace()
+	if scope == meta.RESTScopeNamespace && ns == "" {
+		return field.Required(field.NewPath("metadata", "namespace"), "namespace is required")
+	}
+	if scope == meta.RESTScopeRoot && ns != "" {
+		return field.Invalid(field.NewPath("metadata", "namespace"), ns, "namespace must be empty")
+	}
+	return nil
+}
+
+func (v *Validator) validateDependsOnAnnotation(obj *unstructured.Unstructured, ids object.ObjMetadataSet) []error {
+	if !dependson.HasAnnotation(obj) {
+		return nil
+	}
+
+	deps, err := dependson.ReadAnnotation(obj)
+	if err != nil {
+		return []error{err}
+	}
+
+	id := UnstructuredToObjMetadata(obj)
+	var errors []error
+
+	// duplicates are not allowed
```

Comment:
Why duplicates are not allowed here?

Reply:
If the user put two of the same dependencies in the same depends-on annotation, it's probably a mistake, right?

Reply:
I've moved this to SortObjs and comments on the why,.